src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowJob.java

```diff
@@ -339,12 +343,29 @@ public String getShortDescription() {
 
     @Exported
     @Override public boolean isConcurrentBuild() {
-        return !Boolean.FALSE.equals(concurrentBuild);
+        return getProperty(DisableConcurrentBuildsJobProperty.class) == null;
     }
-    
+
     public void setConcurrentBuild(boolean b) throws IOException {
-        concurrentBuild = b ? null : false;
-        save();
+        concurrentBuild = null;
+
+        boolean propertyExists = getProperty(DisableConcurrentBuildsJobProperty.class) != null;
+
+        // If the property exists, concurrent builds are disabled. So if the argument here is true and the
+        // property exists, we need to remove the property, while if the argument is false and the property
+        // does not exist, we need to add the property. Yay for flipping boolean values around!
+        if (propertyExists == b) {
+            BulkChange bc = new BulkChange(this);
+            try {
+                removeProperty(DisableConcurrentBuildsJobProperty.class);
+                if (!b) {
+                    addProperty(new DisableConcurrentBuildsJobProperty());
+                }
+                bc.commit();
+            } finally {
+                bc.abort();
```

Comment:
Ah, this was the paradigm used before https://github.com/jenkinsci/jenkins/commit/5b4dbb3970b7e559e32aee6dbd159e66bf7a68a8, and since we're not basing on HEAD of core, we still need to do things the old way.